libflux/src/semantic/env.rs

```diff
@@ -0,0 +1,66 @@
+use crate::semantic::sub::{Subst, Substitutable};
+use crate::semantic::types::PolyType;
+use crate::semantic::vars::Tvars;
+use std::collections::HashMap;
+
+// A type environment maps program identifiers to their generic types
+pub struct Environment {
+    parent: Option<Box<Environment>>,
+    values: HashMap<String, PolyType>,
+}
+
+impl Substitutable for Environment {
+    fn apply(self, sub: &Subst) -> Self {
+        let values = self.values.apply(sub);
+        let parent = if let Some(env) = self.parent {
+            Some(Box::new(env.apply(sub)))
+        } else {
+            None
+        };
+        Environment { parent, values }
+    }
+    fn free_vars(&self) -> Tvars {
+        let vars = self.values.free_vars();
+        if let Some(env) = &self.parent {
+            vars.union(env.free_vars())
+        } else {
+            vars
+        }
+    }
+}
+
+impl Environment {
+    pub fn new() -> Environment {
+        Environment {
+            parent: None,
+            values: HashMap::new(),
+        }
+    }
+    pub fn from(parent: Self) -> Environment {
+        Environment {
+            parent: Some(Box::new(parent)),
+            values: HashMap::new(),
+        }
+    }
+    pub fn lookup(&self, v: &str) -> Option<PolyType> {
+        if let Some(t) = self.values.get(v) {
+            Some(t.clone())
+        } else {
+            if let Some(env) = &self.parent {
+                env.lookup(v)
+            } else {
+                None
+            }
+        }
+    }
+    pub fn add(&mut self, name: String, t: PolyType) {
+        self.values.insert(name, t);
+    }
+    pub fn pop(self) -> Environment {
```

Comment:
Not sure about this one. I will definitely add a comment, but `pop` should never be called on a top-level environment. That is a type environment that isn't nested inside of another. That's why I didn't have it return an option. However I think I should change it so that if ever an attempt is made to pop the last stack frame.

Reply:
How does a caller know if the env they have is the top level or not?

Reply:
They don't. Meaning that a caller cannot call `pop` without calling `Environment::from()` before it.

Reply:
ok, can you document that then? State that `pop` must not be called unless the caller has also called `from`

Reply:
I was thinking panic as there isn't really a way to handle that. It should just never happen.